Kernel/FileSystem/ISO9660FileSystem.cpp

```diff
@@ -0,0 +1,695 @@
+    if (!block) {
+        return ENOMEM;
+    }
+    auto block_buffer = UserOrKernelBuffer::for_kernel_buffer(block->data());
+
+    size_t total_bytes = min(size, data_length - offset);
+    size_t nread = 0;
+    size_t blocks_already_read = offset / file_system.m_logical_block_size;
+    size_t initial_offset = offset % file_system.m_logical_block_size;
+
+    auto current_block_index = BlockBasedFileSystem::BlockIndex { extent_location + blocks_already_read };
+    while (nread != total_bytes) {
+        size_t bytes_to_read = min(total_bytes - nread, file_system.logical_block_size() - initial_offset);
+        auto buffer_offset = buffer.offset(nread);
+        dbgln_if(ISO9660_VERY_DEBUG, "ISO9660Inode::read_bytes: Reading {} bytes into buffer offset {}/{}, logical block index: {}", bytes_to_read, nread, total_bytes, current_block_index.value());
+
+        if (bool result = file_system.raw_read(current_block_index, block_buffer); !result) {
+            return EIO;
+        }
+
+        bool result = buffer_offset.write(block->data() + initial_offset, bytes_to_read);
+        if (!result) {
+            return EFAULT;
+        }
+
+        nread += bytes_to_read;
+        initial_offset = 0;
+        current_block_index = BlockBasedFileSystem::BlockIndex { current_block_index.value() + 1 };
+    }
+
+    return nread;
+}
+
+InodeMetadata ISO9660Inode::metadata() const
+{
+    return m_metadata;
+}
+
+KResult ISO9660Inode::traverse_as_directory(Function<bool(FileSystem::DirectoryEntryView const&)> visitor) const
+{
+    auto& file_system = static_cast<ISO9660FS const&>(fs());
+    Array<u8, max_file_identifier_length> file_identifier_buffer;
+
+    auto traversal_result = file_system.visit_directory_record(&m_record, [&](ISO::DirectoryRecordHeader const* record) {
+        StringView filename = get_normalized_filename(record, file_identifier_buffer);
+        dbgln_if(ISO9660_VERY_DEBUG, "traverse_as_directory(): Found {}", filename);
+
+        InodeIdentifier id { fsid(), get_inode_index(record, filename) };
+        auto entry = FileSystem::DirectoryEntryView(filename, id, static_cast<u8>(record->file_flags));
+
+        if (!visitor(entry)) {
+            return RecursionDecision::Break;
+        }
+
+        return RecursionDecision::Continue;
+    });
+
+    if (traversal_result.is_error())
+        return traversal_result;
+
+    return KSuccess;
+}
+
+RefPtr<Inode> ISO9660Inode::lookup(StringView name)
+{
+    auto& file_system = static_cast<ISO9660FS const&>(fs());
+    RefPtr<Inode> inode;
+    Array<u8, max_file_identifier_length> file_identifier_buffer;
+
+    auto traversal_result = file_system.visit_directory_record(&m_record, [&](ISO::DirectoryRecordHeader const* record) {
+        StringView filename = get_normalized_filename(record, file_identifier_buffer);
+
+        if (filename == name) {
+            inode = ISO9660Inode::create_from_directory_record(const_cast<ISO9660FS&>(file_system), record, filename);
```

Comment:
This allocates and doesn't observe possible failure 😢

Reply:
There's not much I can do here, the `Inode` API needs to be fixed to return a `KResultOr<RefPtr<Inode>>` from `lookup`. I can, however, `dbgln` something and leave the `RefPtr` empty.